spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -687,28 +690,89 @@ public void handleDelivery(String consumerTag, Envelope envelope,
 					if (this.isRabbitTxManager) {
 						ConsumerChannelRegistry.registerConsumerChannel(getChannel(), this.connectionFactory);
 					}
-					new TransactionTemplate(this.transactionManager, this.transactionAttribute).execute(s -> {
+					if (this.transactionTemplate == null) {
+						this.transactionTemplate =
+								new TransactionTemplate(this.transactionManager, this.transactionAttribute);
+					}
+					this.transactionTemplate.execute(s -> {
 						ConnectionFactoryUtils.bindResourceToTransaction(new RabbitResourceHolder(getChannel(), false),
 								this.connectionFactory, true);
-						callExecuteListener(message, deliveryTag);
+						// unbound in ResourceHolderSynchronization.beforeCompletion()
+						try {
+							callExecuteListener(message, deliveryTag);
+						}
+						catch (RuntimeException e1) {
+							throw e1;
+						}
+						catch (Throwable e2) { //NOSONAR ok to catch Throwable here because we re-throw it below
+							throw new WrappedTransactionException(e2);
+						}
 						return null;
 					});
 				}
+				catch (Throwable e) {
+					if (e instanceof WrappedTransactionException) {
+						if (e.getCause() instanceof Error) {
+							throw (Error) e.getCause();
+						}
+					}
+				}
 				finally {
 					if (this.isRabbitTxManager) {
 						ConsumerChannelRegistry.unRegisterConsumerChannel();
 					}
 				}
 			}
 			else {
-				callExecuteListener(message, deliveryTag);
+				try {
+					callExecuteListener(message, deliveryTag);
+				}
+				catch (Exception e) {
+					// NOSONAR
+				}
 			}
 		}
 
-		private void callExecuteListener(Message message, long deliveryTag) {
+		private void callExecuteListener(Message message, long deliveryTag) throws Exception {
+			boolean channelLocallyTransacted = isChannelLocallyTransacted();
 			try {
 				executeListener(getChannel(), message);
-				boolean channelLocallyTransacted = isChannelLocallyTransacted();
+				handleAck(deliveryTag, channelLocallyTransacted);
+			}
+			catch (ImmediateAcknowledgeAmqpException e) {
+				if (this.logger.isDebugEnabled()) {
+					this.logger.debug("User requested ack for failed delivery");
+				}
+				handleAck(deliveryTag, channelLocallyTransacted);
+			}
+			catch (Exception e) {
+				if (causeChainHasImmediateAcknowledgeAmqpException(e)) {
+					if (this.logger.isDebugEnabled()) {
+						this.logger.debug("User requested ack for failed delivery");
+					}
+				}
+				else {
+					this.logger.error("Failed to invoke listener", e);
+					if (this.transactionManager != null) {
+						if (this.transactionAttribute.rollbackOn(e)) {
+							rollback(deliveryTag, e);
+							throw e;
+						}
+						else {
+							if (this.logger.isDebugEnabled()) {
+								this.logger.debug("No rollback for " + e);
+							}
+						}
+					}
+					else {
+						rollback(deliveryTag, e);
+					}
```

Comment:
In this else block, we're not running in a transaction template, so there is no reason to throw the exception since it is ignored anyway (because we're running on the amqp-client thread). Will add a comment.